### PR TITLE
[FEATURE] - Allow user to set Quota limit parameters from the PVC request

### PR DESCRIPTION
**Pull request by shefali-malhotra**
# Description
- Helm related changes to add  csi-metadata-retriever sidecar support in driver .
- Adding support to configure various Quota Limit parameters from pvc yaml file.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
|https://github.com/dell/csm/issues/742 |

# Checklist:

- [X] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [X] I have verified that new and existing unit tests pass locally with my changes
- [X] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 90% code coverage
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] Backward compatibility is not broken

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [X] Driver installation with metadata side car 
image

- [X] Dev testing performed
- If value is passed only in sc.yaml (soft limit -40 and soft grace period-86400)
- if value is passed only in pvc.yaml ( advisory parameter- 50)
- if value is passed in sc and pvc than pvc should get precedence (soft limit-10 and soft grace period-85400) 
image
image

### CODE REVIEW

File: provider/provider.go [hunk 1]
```diff
@@ -41,10 +42,17 @@ func New() gocsi.StoragePluginProvider {
 	serverOptions[0] = maxStreams
 
 	svc := service.New()
+
+	log.Info("***Shefali: provider.go**")
```
Comment (HarishH-DELL) [inline provider/provider.go:46]: remove the `log.Info("***Shefali: provider.go**")`

File: service/controller.go [hunk 2]
```diff
@@ -19,6 +19,7 @@ package service
 import (
 	"errors"
 	"fmt"
+	//log "github.com/sirupsen/logrus"
```
Comment (HarishH-DELL) [inline service/controller.go:22]: remove the unused import

File: service/mock/k8s/fakeNodeCreator.go [hunk 3]
```diff
@@ -136,8 +136,8 @@ func GetFakeNode() *v1.Node {
 			Annotations:                nil,
 			OwnerReferences:            nil,
 			Finalizers:                 nil,
-			ClusterName:                "",
-			ManagedFields:              nil,
+			//ClusterName:                "",
```
Comment (chimanjain) [inline service/mock/k8s/fakeNodeCreator.go:139]: You can remove this comment.

File: service/interceptor/interceptor.go [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+
+		if metadataRes != nil {
+			for k, v := range metadataRes.Parameters {
+				log.Info("***Start: Meta Data From PVC Yaml File**")
+				log.Info(k)
+				log.Info(v)
+				log.Info("***End: Meta Data From PVC Yaml File**")
+				req.Parameters[k] = v
+			}
+		} else {
```
Comment (HarishH-DELL) [inline service/interceptor/interceptor.go:221]: else condition is empty , please check this

File: service/interceptor/interceptor.go [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	if closer, ok := lock.(io.Closer); ok {
+		defer closer.Close() // #nosec G307
+	}
+
+	if !lock.TryLock(i.opts.timeout) {
+		return nil, status.Error(codes.Aborted, pending)
```
Comment (HarishH-DELL) [inline service/interceptor/interceptor.go:161]: same as above comment

File: service/interceptor/interceptor.go [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+	defer lock.Unlock()
+
+	return handler(ctx, req)
+}
+
+func (i *interceptor) nodeUnstageVolume(ctx context.Context, req *csi.NodeUnstageVolumeRequest,
+	info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (res interface{}, resErr error) {
+	lock, err := i.opts.locker.GetLockWithID(ctx, req.VolumeId)
+	if err != nil {
+		return nil, err
```
Comment (HarishH-DELL) [inline service/interceptor/interceptor.go:172]: please check this it should return the err only according to your condition , but you are returning the nil and err
any specific reason ?

File: service/interceptor/interceptor.go [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+
+	return handler(ctx, req)
+}
+
+func (i *interceptor) createVolume(ctx context.Context, req *csi.CreateVolumeRequest,
+	info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (res interface{}, resErr error) {
+
+	lock, err := i.opts.locker.GetLockWithID(ctx, req.Name)
+	if err != nil {
+		return nil, err
```
Comment (HarishH-DELL) [inline service/interceptor/interceptor.go:191]: please check this it should return the err only according to your condition , but you are returning the nil and err
any specific reason ?

File: service/interceptor/interceptor.go [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+	}
+
+	if closer, ok := lock.(io.Closer); ok {
+		defer closer.Close() // #nosec G307
+	}
+
+	if !lock.TryLock(i.opts.timeout) {
+		return nil, status.Error(codes.Aborted, pending)
```
Comment (HarishH-DELL) [inline service/interceptor/interceptor.go:199]: same as above comment

File: service/interceptor/interceptor.go [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+	metadataReq := &retriever.GetPVCLabelsRequest{
+		Name:      req.Parameters["csi.storage.k8s.io/pvc/pvcsoftlimit"],
+		NameSpace: req.Parameters["csi.storage.k8s.io/pvc/pvcsoftlimit"],
+		// This should be corrected to SoftLimt changes.
+	}
+
+	if i.opts.MetadataSidecarClient != nil {
+		metadataRes, err := i.opts.MetadataSidecarClient.GetPVCLabels(ctx, metadataReq)
+		if err != nil {
+			/*			log.Errorf("Cannot retrieve labels for PVC %s in namespace: %s, error: %v",
```
[the remark this answers is not in the document]
Comment (HarishH-DELL) [reply service/interceptor/interceptor.go]: no executable code inside the if condition please check this

**Review by HarishH-DELL — CHANGES_REQUESTED** [review]

File: service/interceptor/interceptor.go [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+			log.Error("Cannot get csi-metadata-retriever client")
+		}
+
+		i.opts.MetadataSidecarClient = retrieverClient
+	} else {
+		log.Warn("env var not found: ", "CSI_RETRIEVER_ENDPOINT")
+	}
+}
+
+const pending = "pending"
```
[the remark this answers is not in the document]
Comment (shefali-malhotra) [reply service/interceptor/interceptor.go:147]: unexported constant follow lowerCamelCase

File: service/interceptor/interceptor.go [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+
+func (i *interceptor) nodeStageVolume(ctx context.Context, req *csi.NodeStageVolumeRequest,
+	info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (res interface{}, resErr error) {
+	lock, err := i.opts.locker.GetLockWithID(ctx, req.VolumeId)
+	if err != nil {
+		return nil, err
```
[the remark this answers is not in the document]
Comment (shefali-malhotra) [reply service/interceptor/interceptor.go:153]: if you will see the implementation no action is being taken in this function call hence returning nil . This is just pasing controller frm interceptor.go to controller.go

File: service/interceptor/interceptor.go (hunk 4 above)
Comment (shefali-malhotra) [reply service/interceptor/interceptor.go:221]: it is logging a warning. not empty

File: service/interceptor/interceptor.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (shefali-malhotra) [reply service/interceptor/interceptor.go]: please review latest file

**Review by HarishH-DELL — APPROVED** [review]

**Review by nitesh3108 — APPROVED** [review]

File: service/controller.go [hunk 12]
```diff
@@ -332,6 +350,12 @@ func (s *service) CreateVolume(
 		// use the default if not set in the storage class
 		softGracePrd = SoftGracePrdParamDefault
 	}
+	// If value is passed in pvc than it should get precedence
```
Comment (rajkumar-palani) [inline service/controller.go:353]: Pls create a new method by passing "params" and return the needed variables (after assigning values).

File: service/controller.go [hunk 13]
```diff
@@ -297,17 +304,23 @@ func (s *service) CreateVolume(
...
 			softLimit = SoftLimitParamDefault
 		} else {
 			softLimit = params[SoftLimitParam]
 		}
 	} else {
-		// use the default if not set in the storage class
+		// use the default if not set  in the storage class
 		softLimit = SoftLimitParamDefault
 	}
+	// If value is passed in pvc than it should get precedence
```
Comment (rajkumar-palani) [inline service/controller.go:318]: Pls create a new method by passing "params" and return the needed variables (after assigning values).

**Review by rajkumar-palani — CHANGES_REQUESTED** [review]
can you please check the comments?

**Review by rajkumar-palani — CHANGES_REQUESTED** [review]
Can you please check the **Unit Test and Code Coverage** number is deprecated? please make sure it is >=80%

**Review by HarishH-DELL — APPROVED** [review]

File: service/controller.go (hunk 12 above)
Comment (shefali-malhotra) [reply service/controller.go:353]: Will be addressed in upcoming PR

**Review by rajkumar-palani — APPROVED** [review]
lgtm

**Review by chimanjain — APPROVED** [review]

**Comment by chimanjain** [comment]
lgtm

**Pull request merged**